File: Leetcode/src/com/leetcode/greedy/_763_PartitionLabels.java
package com.leetcode.greedy;

import java.util.ArrayList;
import java.util.List;

/**
 * @Date: 05/12/2020
 * @Description: Greedy, Index Map
 **/
public class _763_PartitionLabels {

    public List<Integer> partitionLabels(String S) {
        if (S == null || S.length() == 0) {
            return new ArrayList<>();
        }
        // greedy
        List<Integer> res = new ArrayList<>();
        int[] index = new int[26];
        int n = S.length();
        for (int i = 0; i < n; i++) {
            index[S.charAt(i) - 'a'] = i;
        }
        int begin = 0, end = 0;
        for (int i = 0; i < n; i++) {
            char ch = S.charAt(i);
            end = Math.max(end, index[ch - 'a']);
            if (end == i) {
                // add
                res.add(end - begin + 1);
                begin = i + 1;
            }
        }
        return res;
    }
}

File: Leetcode/src/com/leetcode/greedy/_763_PartitionLabels.py
'''
Date: 11/12/2020 15:29:19
LastEditTime: 11/17/2020 16:30:22
Description: HashMap(dict)
'''
class Solution:
    def partitionLabels(self, S: str) -> List[int]:
        n = len(S)
        # Use HashMap to store the latest index for each char.
        right_most = {ch:index for index, ch in enumerate(S)}

        start, max_pos = 0, 0
        res = []
        for i in range(n):
            max_pos = max(max_pos, right_most[S[i]])
            if i == max_pos:
                res.append(i - start + 1)
                start = i + 1
        return res